### PR TITLE
Manual ack, status check

### PR DESCRIPTION
**Pull request by poiuytrez**
I have added the support for manual ack: 
https://www.rabbitmq.com/confirms.html

A way to get the connection status (#5)

I removed also the support for groovy as I won't be able to maintain it. It does not compile and I don't know how to fix it.

### CODE REVIEW

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 1]
```diff
@@ -1,16 +1,14 @@
 package io.vertx.rabbitmq;
 
-import io.vertx.codegen.annotations.VertxGen;
 import io.vertx.core.AsyncResult;
 import io.vertx.core.Handler;
 import io.vertx.core.Vertx;
 import io.vertx.core.json.JsonObject;
-import io.vertx.rabbitmq.impl.RabbitMQClientImpl;
+import io.vertx.rabbitmq.impl.RabbitMQClientImpl;import java.lang.String;import java.lang.Void;
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: looks like there are missing \n

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 2]
```diff
@@ -20,12 +18,26 @@ static RabbitMQClient create(Vertx vertx, JsonObject config) {
   //TODO: Think about splitting this out into different API's with specific roles (admin, pub, sub)
   //TODO: Simplify/Change name of API methods to match more vert.x type verbiage ?
 
+  void basicAck(long deliveryTag, boolean multiple, Handler<AsyncResult<JsonObject>> resultHandler);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: javadoc

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 3]
```diff
@@ -20,12 +18,26 @@ static RabbitMQClient create(Vertx vertx, JsonObject config) {
...
+
+  void basicNack(long deliveryTag, boolean multiple, boolean requeue, Handler<AsyncResult<JsonObject>> resultHandler);
```
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: javadoc

**Comment by poiuytrez** [comment]
Done

File: src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java [hunk 4]
```diff
@@ -46,9 +40,46 @@ public RabbitMQClientImpl(Vertx vertx, JsonObject config) {
...
+  @Override
   public void basicConsume(String queue, String address, Handler<AsyncResult<Void>> resultHandler) {
+    basicConsume(queue, address, true, resultHandler);
+  }
+
+  @Override
+  public void basicConsume(String queue, String address, boolean autoAck, Handler<AsyncResult<Void>> resultHandler) {
     forChannel(resultHandler, channel -> {
-      channel.basicConsume(queue, new ConsumerHandler(vertx, channel, includeProperties, ar -> {
+      channel.basicConsume(queue, new ConsumerHandler(vertx, channel, includeProperties, autoAck, ar -> {
```
Comment (agriesser) [inline src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java]: Hm shouldn't we call basicConsume(queue, true, new ConsumerHandler(....)); in case we want to have autoAck? This is more performant since the rabbitmq broker doesn't have to wait for any ACKs received from this consumer (helps with throughput). In that case the ConsumerHandler obviously has to refrain from acking the message.

**Comment by diega** [comment]
Doesn't this work duplicate part of [#6](https://github.com/vert-x3/vertx-rabbitmq-client/pull/6)?

File: src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java (hunk 4 above)
Comment (poiuytrez) [inline src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java]: done.

**Comment by poiuytrez** [comment]
@diega it's a duplicate. I just closed it.

File: src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java [hunk 5]
```diff
@@ -46,9 +40,46 @@ public RabbitMQClientImpl(Vertx vertx, JsonObject config) {
...
+  public void basicNack(long deliveryTag, boolean multiple, boolean requeue, Handler<AsyncResult<JsonObject>> resultHandler) {
+    forChannel(resultHandler, (channel) -> {
+      channel.basicNack(deliveryTag, multiple, requeue);
+      return null;
+    });
+  }
+
+  @Override
   public void basicConsume(String queue, String address, Handler<AsyncResult<Void>> resultHandler) {
+    basicConsume(queue, address, true, resultHandler);
```
Comment (agriesser) [inline src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java]: The docs you added in the interface specify that the basicConsume(...) method with 3 parameters produces an explicit acknowledgement consumer. Therefore the 3. parameter should be false here, shouldn't it?

Comment (poiuytrez) [inline src/main/java/io/vertx/rabbitmq/impl/RabbitMQClientImpl.java]: You are right. It is an error in the docs. I've changed the docs.

File: README.md [hunk 6]
```diff
@@ -94,6 +94,28 @@ client.basicGet("my.queue", true, getResult -> {
 });
 ```
 
+## Consume messages without auto-ack
```
Comment (cescoffier) [inline README.md]: At some point we would need to switch to the "way" vert.x stack is handling documentation.

File: pom.xml [hunk 7]
```diff
@@ -12,10 +12,10 @@
   <modelVersion>4.0.0</modelVersion>
 
   <artifactId>vertx-rabbitmq-client</artifactId>
-  <version>3.0.0-SNAPSHOT</version>
+  <version>3.2.1</version>
```
Comment (cescoffier) [inline pom.xml]: Must be a SNAPSHOT.

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 8]
```diff
@@ -1,16 +1,18 @@
...
 import io.vertx.core.json.JsonObject;
 import io.vertx.rabbitmq.impl.RabbitMQClientImpl;
+import java.lang.String;
+import java.lang.Void;
 
 /**
  * @author <a href="mailto:nscavell@redhat.com">Nick Scavelli</a>
  */
-@VertxGen
 public interface RabbitMQClient {
```
Comment (cescoffier) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: Why did you remove the @VertxGen annotation ?

**Comment by cescoffier** [comment]
It looks like the polyglot generation has been removed ? Such kind of client should be polyglot I think, especially, RabbitMQ can be used from multiple languages, so I don't see why we would not be able to do it. 

_Note related to this issue_: At some point we can see if we want to build a bridge between the event bus and rabbitMQ.

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 9]
```diff
@@ -20,12 +22,52 @@ static RabbitMQClient create(Vertx vertx, JsonObject config) {
...
+   * Request specific "quality of service" settings, Limiting the number of unacknowledged messages on
+   * a channel (or connection). This limit is applied separately to each new consumer on the channel.
+   *
+   * @see com.rabbitmq.client.Channel#basicQos(int)
+   */
+  void basicQos(int prefetchCount, Handler<AsyncResult<Void>> resultHandler);
+
   void exchangeDeclare(String exchange, String type, boolean durable, boolean autoDelete, Handler<AsyncResult<Void>> resultHandler);
 
   void exchangeDelete(String exchange, Handler<AsyncResult<Void>> resultHandler);
```
Comment (cescoffier) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: Documentation missing

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 10]
```diff
@@ -73,4 +115,18 @@ static RabbitMQClient create(Vertx vertx, JsonObject config) {
   void start(Handler<AsyncResult<Void>> resultHandler);
```
Comment (cescoffier) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: Documentation missing.

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java [hunk 11]
```diff
@@ -73,4 +115,18 @@ static RabbitMQClient create(Vertx vertx, JsonObject config) {
...
 
   void stop(Handler<AsyncResult<Void>> resultHandler);
```
Comment (cescoffier) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: Documentation missing.

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 8 above)
Comment (vietj) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: we need to keep the polyglot aspect, it's one of the key features of vertx

File: src/main/java/io/vertx/rabbitmq/RabbitMQClient.java (hunk 10 above)
Comment (poiuytrez) [inline src/main/java/io/vertx/rabbitmq/RabbitMQClient.java]: done

**Comment by poiuytrez** [comment]
Could you check the PR again? Thank you.

**Comment by cescoffier** [comment]
I wonder why the polyglot files are not generated.

**Comment by vietj** [comment]
the lang-js and lang-groovy dependencies have been removed from the pom.xml

**Comment by poiuytrez** [comment]
I have added back vertx-lang-groovy but the groovy code is not generated.

**Comment by vietj** [comment]
you need to extend the vertx-ext-parent, look at for example https://github.com/vert-x3/vertx-jdbc-client/blob/master/pom.xml

**Comment by poiuytrez** [comment]
This time, I have the groovy code.

**Comment by poiuytrez** [comment]
Should I put the generated code in the repo?

**Comment by vietj** [comment]
yes it should be in the repo, look at how other project do

**Comment by vietj** [comment]
check also you have all langs, like RxJava

**Comment by poiuytrez** [comment]
@vietj I have added RxJava.

**Comment by vietj** [comment]
great
